### ISSUE TITLE
Mails not signed with DKIM key if sending host's IP is in "Weiterleitungs-Hosts (WebUI)"

### ISSUE THREAD

**Issue by Adorfer** [mailcow/mailcow-dockerized#1740]
looks a blit like #1091, but here the rebuild/update did not cure the issue. 

config: https://pastebin.com/cfbatu2T
ipv4tables: https://pastebin.com/AfFPuGii
ipv6tables: https://pastebin.com/SeT5jCiS
works for one domain:  https://pastebin.com/UmV4eggU
does not work for other domain: https://pastebin.com/6bHkvTqB

**Comment by andryyy** [comment]
set pre-result to no action: 'Matched map: KEEP_SPAM'

Probably a mail to a FWD host? Any idea, @mkuron?

**Comment by Adorfer** [comment]
the mail is sent via account (with login), but from a host which is in the forward-host-whitlist ("my networks").

**Comment by mkuron** [comment]
Good catch. This happens when an authenticated sender sends _from_ an IP in the forwarding hosts list. It does not happen when sending _to_ a forwarding host. I believe the solution is to add a composite rule to data/conf/local.d/composites.conf that removes the `KEEP_SPAM` symbol if the `MAILCOW_AUTH` symbol is present while keeping the `MAILCOW_AUTH` symbol (that’s what the `-` is for):
```
AUTH_AND_FORWARDING_HOST {
  expression = "-MAILCOW_AUTH & KEEP_SPAM"
}
```

**Comment by Adorfer** [comment]
> data/conf/local.d/composites.conf

"local.d" does not exist in /opt/mailcow-dockerized/data/conf

**Comment by Adorfer** [comment]
I found composites.conf in data/conf/rspamd/local.d/
but addig those 2,5 lines in there and  running the update-procedure did not change a thing. 

"DKIM-Result: none (no signature)"

**Comment by andryyy** [comment]
Writing from mobile: Isn’t it !MAILCOW_AUTH instead of -?

**Comment by mkuron** [comment]
`!` is the logical not operator. `-` is not an operator but means that the symbol should be kept. By default, a composite symbol replaces the symbols it is based on. https://www.rspamd.com/doc/configuration/composites.html

I didn't have the time yet to test this, but I think it should work. Did you restart rspamd, @Adorfer? `docker-compose restart rspamd-mailcow`.

**Comment by andryyy** [comment]
But it should work with ‘!’ - the rule would not be applied then.

It still needs more changes, but it should at least sign messages again. The logic for forward hosts is skipped, yes, but it should work as workaround.

**Comment by Adorfer** [comment]
I did restart. 
But i do not understand your ! - discussion above. 
what should i do exactly?

**Comment by Adorfer** [comment]
Any other suggestion what i could do? 
Do you perhaps need additional debug information? 
The "Dkim not working for that domain" is causing outgoing mails going into spam at some providers.

**Comment by andryyy** [comment]
You tried to replaced „-„ by „!“?

**Comment by Adorfer** [comment]
yes, but i it's not clear for me if i did edit the right file, because the initial statement said, that it's in a different directory.
Or shall i create that directory?
![grafik](https://user-images.githubusercontent.com/10302493/45162104-83b76580-b1ed-11e8-8003-ec01ba9936b3.png)

(and yes, i did rebuild + update + restart the VM a few times during the last 3 days, no change in the outcome, neither with "-", nor with "!". So i do not understand the discussion about the forwarding-hosts. If there is a shortcut like "remove the ipv6 and/ord the users ipv4 from list xy": i would do so. But just explain me exacltly which one. are we talking about the table/list in the admin-ui or the list of fqdns in the config-file? And if the last, which line in which file?)

**Comment by andryyy** [comment]
You don’t need to update, rebuild or restart the VM. Just restart Rspamd after changing the file. ;-)

What do you want to understand about the discussion about the forward hosts? It _is_ the problem. I just had no time to fix this. 

I guess mailcow_auth is called **after** Rspamd set the prefilter result. We just need to check for authentication in a different way. 

I will check this when you grant me access again. It is easier than setting up that exact scenario in a virtual environment. 

The file was data/conf/rspamd/local.d, mkuron just missed “rspamd” in the path. 

“Just“ explaining everything with every line etc. is not possible.... :-) My time is limited.

**Comment by Adorfer** [comment]
I just removed the IP from the list, now it's signed successfully. 
![grafik](https://user-images.githubusercontent.com/10302493/45194715-80a39000-b254-11e8-82fa-1988bdbd7e16.png)

![grafik](https://user-images.githubusercontent.com/10302493/45194729-93b66000-b254-11e8-90ab-a1fbb0c89f8b.png)

(For me it can be closed here. please note this side effect in the docs.)

**Comment by mkuron** [comment]
We still need to fix this, so please leave the issue open. I just didn’t get to it yet.

**Comment by andryyy** [comment]
Could work with

```
KEEP_SPAM {
  type = "ip";
  map = "redis://KEEP_SPAM";
  action = "accept";
  symbols_set = ["KEEP_SPAM"];
  nflags = ["authenticated"];
}
```

in multimap.

**Comment by mkuron** [comment]
So the composite rule definitely doesn't work because the KEEP_SPAM rule accepts pre-filter, so the composite rule is never run. `nflags` in multimap looks like it should work -- but it doesn't for some reason. I'll need to do some debugging, perhaps the flags aren't yet set at this point.

**Comment by mkuron** [comment]
`nflags` only applies to multimaps with `type = "received"`, so it's not going to help us here. `require_symbols` would do, but it can't be used in `prefilter` mode (i.e. when an `action` is specified).

**Comment by mkuron** [comment]
@Adorfer, I found a solution. Could you put the IP back into the list and then try out the change from #1748?

**Comment by andryyy** [comment]
We could write it in LUA. Just check against the map and set the pre result if get_user is false.

**Comment by mkuron** [comment]
@andryyy, remove the `action` from `KEEP_SPAM` and write a custom pre-filter rule that depends on `KEEP_SPAM` and removes that symbol if `get_user` is true or sets the action if false? Sounds like a plan.

**Comment by andryyy** [comment]
We can also query the map from within lua. :-) Sounds like it should work.

**Comment by Adorfer** [comment]
@mkuron PR from #1748 works.

**Comment by andryyy** [comment]
It does, yes. But see the PR comments. 

I wrote a function for it yesterday. Will push some more changes today.

**Comment by Adorfer** [comment]
sorry for answering the question as i did understand it.

**Comment by andryyy** [comment]
Hm?

**Comment by mkuron** [comment]
Fixed by afc18fd46935e201a9be85ab636f0ed0b77209be
